### PR TITLE
Feat/update runtime v0.9

### PR DESCRIPTION
**Pull request by ZenGround0**
Update v0.9 runtime to the agreed refactor
- [x] Inline sub interfaces
- [x] Modify send to have an output parameter
- [x] Tracing sub interface becomes single function
- [x] Extract log and cbor types importing latest go-state-types

Still WIP.  The inlining is done and all of the send modification for source.  Tests for send need fixing along with the last two items.

### CODE REVIEW

**Comment by ZenGround0** [comment]
@anorth I know we won't merge this yet, but its ready for review.  Planning to cherrypick this over to master once review passes.

File: actors/runtime/runtime.go [hunk 1]
```diff
@@ -141,8 +125,8 @@ type Runtime interface {
 	// should not be used by actor code directly.
 	Context() context.Context
 
-	// Starts a new tracing span. The span must be End()ed explicitly, typically with a deferred invocation.
-	StartSpan(name string) TraceSpan
+	// Starts a new tracing span. The span must be End()ed explicitly by invoking or deferring the returned end function
+	StartSpan(name string) func()
```
Comment (anorth) [inline actors/runtime/runtime.go:129]: Can you use named return value syntax to name it `EndSpan`?

File: actors/runtime/runtime.go [hunk 2]
```diff
@@ -152,15 +136,15 @@ type Runtime interface {
...
 	// Note events that may make debugging easier
-	Log(level LogLevel, msg string, args ...interface{})
+	Log(level rt.LogLevel, msg string, args ...interface{})
 }
 
 // Store defines the storage module exposed to actors.
 type Store interface {
 	// Retrieves and deserializes an object from the store into `o`. Returns whether successful.
-	Get(c cid.Cid, o CBORUnmarshaler) bool
+	Get(c cid.Cid, o cbor.Unmarshaler) bool
```
Comment (anorth) [inline actors/runtime/runtime.go:145]: As embeds, I think these need longer names, `StoreGet` and `StorePut`.

File: actors/runtime/runtime.go [hunk 3]
```diff
@@ -213,25 +197,19 @@ type Syscalls interface {
 // the return, in particular whether it has been serialized to bytes or just passed through.
 // Production code is expected to de/serialize, but test and other code may pass the value straight through.
 type SendReturn interface {
```
Comment (anorth) [inline actors/runtime/runtime.go:199]: Delete SendReturn?

File: actors/runtime/runtime.go [hunk 4]
```diff
@@ -213,25 +197,19 @@ type Syscalls interface {
...
-	End()
+	Into(cbor.Unmarshaler) error
 }
 
 // StateHandle provides mutable, exclusive access to actor state.
 type StateHandle interface {
 	// Create initializes the state object.
 	// This is only valid in a constructor function and when the state has not yet been initialized.
-	Create(obj CBORMarshaler)
+	Create(obj cbor.Marshaler)
```
Comment (anorth) [inline actors/runtime/runtime.go:207]: Similar for naming: StateCreate, StateRead, StateTransact

The Message and Syscalls embeds are fine without longer names, though.

File: actors/runtime/runtime.go [hunk 5]
```diff
@@ -305,3 +268,16 @@ func (b *CBORBytes) UnmarshalCBOR(r io.Reader) error {
 	*b = c.Bytes()
 	return err
 }
+
+// Discard is a helper
+type Discard struct{}
```
Comment (anorth) [inline actors/runtime/runtime.go:273]: This is now an internal detail for actors, not about their interface to the VM. Note there are no more concrete types in this file.

Move to `builtin/shared.go` or similar.

**Review by anorth — APPROVED** [review]
Great work.

I'm not sure that removing the puppet actor is safe in this branch. Could you check with test-vectors team? I dunno if chain-validation is still in use in lotus master.

**Comment by ZenGround0** [comment]
> I'm not sure that removing the puppet actor is safe in this branch. Could you check with test-vectors team? I dunno if chain-validation is still in use in lotus master.

Yup its off lotus master a few days back [PR here](https://github.com/filecoin-project/lotus/pull/3606/commits)

**Pull request merged**